bridges/centralized-ethereum/src/actors/dr_reporter.rs

```diff
@@ -88,116 +96,175 @@ impl Handler<DrReporterMsg> for DrReporter {
+                } else {
+                    reports.push(report);
+                }
+            }
+            msg.reports = reports;
+
+            if msg.reports.is_empty() {
+                // Nothing to report
                 return;
             }
 
-            // Report result
-            let dr_gas_price: Result<U256, web3::contract::Error> = wrb_contract
-                .query(
-                    "readRequestGasPrice",
-                    msg.dr_id,
-                    eth_account,
-                    contract::Options::default(),
-                    None,
-                )
-                .await;
-
-            match dr_gas_price {
-                Ok(dr_gas_price) => {
-                    log::debug!("Request [{}], calling reportResult", msg.dr_id);
-                    // FIXME(#2046): ReportResult with 4 arguments (with timestamp)
+            let max_gas_price = get_max_gas_price(&msg, &wrb_contract, eth_account).await;
+
+            if max_gas_price == U256::from(0u8) {
+                // Error reading gas price, abort
+                return;
+            }
+
+            log::debug!("Request [{:?}], calling reportResultBatch", dr_ids);
+            let batch_results: Vec<_> = msg
+                .reports
+                .iter()
+                .map(|report| {
+                    let dr_hash = H256::from_slice(report.dr_tx_hash.as_ref());
+
+                    // the trait `web3::contract::tokens::Tokenize` is not implemented for
+                    // `(std::vec::Vec<(web3::types::U256, web3::types::U256, web3::types::H256, std::vec::Vec<u8>)>, bool)
+                    // Need to manually convert to tuple
+                    Token::Tuple(vec![
+                        Token::Uint(report.dr_id),
+                        Token::Uint(report.timestamp.into()),
+                        Token::FixedBytes(dr_hash.to_fixed_bytes().to_vec()),
+                        Token::Bytes(report.result.clone()),
+                    ])
+                })
+                .collect();
+            let verbose = true;
+            let params_str = format!("{:?}", (&batch_results, verbose));
+
+            let batches = split_by_gas_limit(
+                batch_results,
+                &wrb_contract,
+                eth_account,
+                report_result_limit,
+                verbose,
+                max_gas_price,
+            )
+            .await;
+
+            for (batch_results, estimated_gas_limit) in batches {
+                let only_1_batch = batch_results.len() == 1;
+                let receipt = if only_1_batch {
+                    let (dr_id, ts, dr_tx_hash, report_result) =
+                        unwrap_batch(batch_results[0].clone());
+
                     let receipt_fut = wrb_contract.call_with_confirmations(
                         "reportResult",
-                        (msg.dr_id, dr_hash, msg.result),
+                        (dr_id, ts, dr_tx_hash, report_result),
```

Comment:
Maybe this works?

```suggestion
batch_results[0].clone(),
```

Reply:
at least compile...

Reply:
Unfortunately it didn't work, because when a function has 4 arguments, the web3 library expects a tuple with 4 elements. And batch_results[0] is a Token::Tuple with 4 elements, but that doesn't count as 4 arguments but as 1.